### PR TITLE
PodSecurityContext with inline fields

### PR DESCRIPTION
**Pull request by pmorie**
... as proposed in #12823.  Based upon the skeletal commit in #13939, which is the basis for other PRs soon to land.  I am opening a new PR for this so that we can easily rebase other PRs which depend on the initial commit in #13939.

### CODE REVIEW

File: pkg/securitycontext/provider.go [hunk 1]
```diff
@@ -37,10 +37,8 @@ type SimpleSecurityContextProvider struct{}
 // The security context provider can make changes to the Config with which
 // the container is created.
 func (p SimpleSecurityContextProvider) ModifyContainerConfig(pod *api.Pod, container *api.Container, config *docker.Config) {
-	if container.SecurityContext == nil {
-		return
-	}
-	if container.SecurityContext.RunAsUser != nil {
+	effectiveSC := determineEffectiveSecurityContext(pod, container)
```
Comment (pmorie) [inline pkg/securitycontext/provider.go]: @pweil- Particularly interested in your thoughts on this code before I write a bunch of tests for code that's wrong :-D

File: pkg/securitycontext/provider.go [hunk 2]
```diff
@@ -95,3 +96,53 @@ func makeCapabilites(capAdd []api.Capability, capDrop []api.Capability) ([]strin
 	}
 	return addCaps, dropCaps
 }
+
+func determineEffectiveSecurityContext(pod *api.Pod, container *api.Container) *api.SecurityContext {
```
Comment (pmorie) [inline pkg/securitycontext/provider.go]: @pweil- @smarterclayton @liggitt I'm starting to think this is wrong.  I think if you have `container.SecurityContext` specified at all, it should win for that container over `pod.Spec.SecurityContext`, period.  Otherwise old clients cannot reason correctly about what a container's `SecurityContext` is.

**Comment by pmorie** [comment]
cc @wattsteve @childsb @swagiaal

File: pkg/securitycontext/provider.go (hunk 1 above)
Comment (pweil-) [inline pkg/securitycontext/provider.go]: Still need a nil check here if PSC and CSC aren't set but this otherwise matches up with what we were discussing.

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/securitycontext/provider.go]: That is correct.

**Comment by jlowdermilk** [comment]
Reassigning to @thockin who reviewed original proposal. cc @bgrant0607

**Comment by pmorie** [comment]
FYI, this isn't ready for review yet; I need to fix an obvious issue with
the current impl.  Will tag tim in when i'm ready.

**Comment by pmorie** [comment]
This is ready for review.

**Comment by pmorie** [comment]
Rebased on top of the merged #13939

File: pkg/api/types.go [hunk 3]
```diff
@@ -1007,6 +1007,23 @@ type PodSecurityContext struct {
...
+	// SELinuxOptions is the SELinux context to be applied to all containers
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
+
+	// RunAsUser is the UID to run the entrypoint of all container processes
+	RunAsUser *int64 `json:"runAsUser,omitempty"`
+
+	// RunAsNonRoot indicates that all containers should be run as a non-root user.
+	// If the RunAsUser field is not explicitly set then the kubelet may check the
+	// image for a specified user or perform defaulting to specify a user.
+	RunAsNonRoot bool
```
Comment (swagiaal) [inline pkg/api/types.go]: Do we need  `json:"runAsNonRoot,omitempty"` here ?

Comment (pmorie) [inline pkg/api/types.go]: Wow, we do.  It's missing in master.

**Comment by pmorie** [comment]
@deads2k @swagiaal PTAL

**Comment by pmorie** [comment]
Removed WIP tag as I believe this should be ready for final review.

File: pkg/api/types.go [hunk 4]
```diff
@@ -1007,6 +1007,23 @@ type PodSecurityContext struct {
 	// Use the host's ipc namespace.
 	// Optional: Default to false.
 	HostIPC bool `json:"hostIPC,omitempty"`
+
+	// Capabilities are the capabilities to add/drop when running containers
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
+
+	// Run all containers in privileged mode
```
Comment (thockin) [inline pkg/api/types.go]: Document default behavior/values for all fields.

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -1249,6 +1249,22 @@ type PodSpec struct {
 
 // PodSecurityContext holds pod-level security attributes and common container settings.
 type PodSecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running containers
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
+
+	// Run all containers in privileged mode
```
Comment (thockin) [inline pkg/api/v1/types.go]: same - document defaults

File: pkg/api/v1/types.go [hunk 6]
```diff
@@ -1249,6 +1249,22 @@ type PodSpec struct {
...
+	Privileged *bool `json:"privileged,omitempty"`
+
+	// SELinuxOptions is the SELinux context to be applied to all containers
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
+
+	// RunAsUser is the UID to run the entrypoint of all container processes
+	RunAsUser *int64 `json:"runAsUser,omitempty"`
+
+	// RunAsNonRoot indicates that all containers should be run as a non-root user.
+	// If the RunAsUser field is not explicitly set then the kubelet may check the
```
Comment (thockin) [inline pkg/api/v1/types.go]: please also fix comments on the old SecurityContext - default values are all wrongly documented

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (thockin) [inline pkg/securitycontext/provider.go]: I am not sure I follow.  If we assert that non-specified values in container.SC mean "do the right thing" and always have, why would a non-nil CSC with no values in it take precedence over PSC with values?

File: pkg/api/types.go [hunk 7]
```diff
@@ -1007,6 +1007,23 @@ type PodSecurityContext struct {
...
+	Privileged *bool `json:"privileged,omitempty"`
```
Comment (deads2k) [inline pkg/api/types.go]: You force this non-nil in your serialization test.  Does this have to match something?  Is `nil` disallowed`?  The godoc doesn't make it obvious why this shouldn't be `nil`.

File: pkg/api/types.go [hunk 8]
```diff
@@ -1007,6 +1007,23 @@ type PodSecurityContext struct {
 	// Use the host's ipc namespace.
 	// Optional: Default to false.
 	HostIPC bool `json:"hostIPC,omitempty"`
+
+	// Capabilities are the capabilities to add/drop when running containers
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
```
Comment (deads2k) [inline pkg/api/types.go]: This is also force non-nil in the fuzzer.  Same questions as with `Privileged`.

File: pkg/api/types.go [hunk 9]
```diff
@@ -1007,6 +1007,23 @@ type PodSecurityContext struct {
...
+
+	// Run all containers in privileged mode
+	Privileged *bool `json:"privileged,omitempty"`
+
+	// SELinuxOptions is the SELinux context to be applied to all containers
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
+
+	// RunAsUser is the UID to run the entrypoint of all container processes
+	RunAsUser *int64 `json:"runAsUser,omitempty"`
```
Comment (deads2k) [inline pkg/api/types.go]: internal types shouldn't have json tags.

File: pkg/api/types.go (hunk 7 above)
Comment (pmorie) [inline pkg/api/types.go]: I think perhaps that was inadvertent in the fuzzing code for `Privileged`, and we might not need it at all.  I'll check and see if we can just remove it.

File: pkg/api/types.go (hunk 8 above)
Comment (pmorie) [inline pkg/api/types.go]: I'm starting to think we can just remove the fuzzers for these, but I'm going to double check when I re mediate the rest of the feedback.

File: pkg/api/types.go (hunk 9 above)
Comment (pmorie) [inline pkg/api/types.go]: They all do, though.

Comment (deads2k) [inline pkg/api/types.go]: > They all do, though.

https://github.com/kubernetes/kubernetes/issues/3933

File: pkg/api/types.go (hunk 7 above)
Comment (ncdc) [inline pkg/api/types.go]: This should not be a pointer

File: pkg/api/v1/types.go [hunk 10]
```diff
@@ -1249,6 +1249,22 @@ type PodSpec struct {
 
 // PodSecurityContext holds pod-level security attributes and common container settings.
 type PodSecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running containers
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
+
+	// Run all containers in privileged mode
+	Privileged *bool `json:"privileged,omitempty"`
```
Comment (ncdc) [inline pkg/api/v1/types.go]: Not a pointer

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (ncdc) [inline pkg/securitycontext/provider.go]: I think I'm with @thockin on this one.

Comment (pmorie) [inline pkg/securitycontext/provider.go]: a non-nil CSC should take precedence

Comment (pmorie) [inline pkg/securitycontext/provider.go]: It does take precedence, in fact, what am I missing?

Comment (ncdc) [inline pkg/securitycontext/provider.go]: If you have a non-nil but empty CSC, that means none of the values from the PSC apply. e.g. if you have PSC.RunAsUser=1234, you lose that because CSC.RunAsUser=nil wins.

File: plugin/pkg/admission/securitycontext/scdeny/admission.go [hunk 11]
```diff
@@ -57,6 +57,14 @@ func (p *plugin) Admit(a admission.Attributes) (err error) {
 	if !ok {
 		return apierrors.NewBadRequest("Resource was marked with kind Pod but was unable to be converted")
 	}
+	if pod.Spec.SecurityContext != nil {
```
Comment (ncdc) [inline plugin/pkg/admission/securitycontext/scdeny/admission.go]: Please update the Godoc for this function as it is currently incorrect

File: test/e2e/security_context.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("[Skipped] Security Context", func() {
```
Comment (ncdc) [inline test/e2e/security_context.go]: Why [Skipped]?

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (pmorie) [inline pkg/securitycontext/provider.go]: if you have a non empty but nil CSC you did it wrong

Comment (pmorie) [inline pkg/securitycontext/provider.go]: the doc fir this feature should handle this corner case and call it out as something not to do

File: pkg/api/types.go (hunk 9 above)
Comment (pmorie) [inline pkg/api/types.go]: I would prefer to make this consistent with the rest of the file even though we know it is not right. i have a laundry list of stuff in that area I want to handle once I have some bandwidth freed up and will fix that issue and writr a verifier around it.

File: test/e2e/security_context.go (hunk 12 above)
Comment (pmorie) [inline test/e2e/security_context.go]: it won't run on all cloud providers currently because many of then disallow security  context. we can work through this.

File: pkg/api/types.go (hunk 7 above)
Comment (pmorie) [inline pkg/api/types.go]: @ncdc I agree it should not be a pointer, but I would prefer to keep the API consistent with `SecurityContext` and change them both at once after this goes in.

File: pkg/api/types.go (hunk 8 above)
Comment (pmorie) [inline pkg/api/types.go]: @deads2k I wound up modifying the fuzzer to fuzz randomly to nil OR a random bool.

**Comment by pmorie** [comment]
@k8s-bot test this

@thockin @deads2k @ncdc feedback addressed, I will make a follow-up issue to fix the pointers in the API.

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by dchen1107** [comment]
ok to test

**Comment by pmorie** [comment]
Thanks @dchen1107

**Comment by pmorie** [comment]
@thockin Should be ready for final review now.

File: pkg/api/types.go [hunk 13]
```diff
@@ -2035,36 +2053,35 @@ type ComponentStatusList struct {
 // will be populated based on the Container configuration if they are not set.  Defining them on
 // both the Container AND the SecurityContext will result in an error.
 type SecurityContext struct {
-	// Capabilities are the capabilities to add/drop when running the container
+	// Capabilities are the capabilities to add/drop when running containers.
+	// Optional: Defaults to the default set of capabilities granted by the container
```
Comment (thockin) [inline pkg/api/types.go]: I am concerned that all the comments on defaults do not mention PodSecurityContext.  Should they?

File: pkg/securitycontext/provider.go [hunk 14]
```diff
@@ -37,11 +37,12 @@ type SimpleSecurityContextProvider struct{}
...
 func (p SimpleSecurityContextProvider) ModifyContainerConfig(pod *api.Pod, container *api.Container, config *docker.Config) {
-	if container.SecurityContext == nil {
+	effectiveSC := determineEffectiveSecurityContext(pod, container)
+	if effectiveSC == nil {
 		return
 	}
-	if container.SecurityContext.RunAsUser != nil {
-		config.User = strconv.FormatInt(*container.SecurityContext.RunAsUser, 10)
+	if effectiveSC.RunAsUser != nil {
+		config.User = strconv.FormatInt(*effectiveSC.RunAsUser, 10)
```
Comment (thockin) [inline pkg/securitycontext/provider.go]: why FormatInt vs IToA?  Asking, not telling.

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (thockin) [inline pkg/securitycontext/provider.go]: So if I have a CSC `{ RunAsUser: 1234 }`, that nullifies my PSC `{ Capabilities: {...}; SELinux: {...}; /* etc */ }` ??  That still feels wrong

File: pkg/api/types.go (hunk 13 above)
Comment (pmorie) [inline pkg/api/types.go]: that's reasonable. i will add it.

File: pkg/securitycontext/provider.go (hunk 14 above)
Comment (pmorie) [inline pkg/securitycontext/provider.go]: itoa would be better, will change

File: pkg/securitycontext/provider.go (hunk 2 above)
Comment (pmorie) [inline pkg/securitycontext/provider.go]: @thockin It does, but that's the only way this can work and be backwards compatible.

Comment (pmorie) [inline pkg/securitycontext/provider.go]: I would say perhaps in the V2 API we can implement field-level munging into an effective security context, but V1 has to be this way for backward compat.

Comment (thockin) [inline pkg/securitycontext/provider.go]: Hold up.  @smarterclayton I thought we (you) made a string case that every field in container.SecurityContext had "do the right thing if not provided" semantics.  I took a look at it, and (while the comments are obsolete and terrible) that seems to be the case.

In this case DTRT means defer to PSC, doesn't it?  Why do I find myself convincing you when you convinced me?

Comment (pmorie) [inline pkg/securitycontext/provider.go]: Do we need to have a hangout to work through this @smarterclayton @thockin ?  I feel like if we are still not clear on the semantics we need to get on the same page and update the recorded semantics in the proposal, since if there's any confusion, the proposal didn't successfully articulate things clearly.

Comment (pmorie) [inline pkg/securitycontext/provider.go]: Added a comment on #14705 to articulate my thoughts on this

**Comment by pmorie** [comment]
@thockin has raised a comment that is now disconnected from this PR here:
https://github.com/kubernetes/kubernetes/pull/14705/files#r42267917, bringing it here so we can move
the discussion to the actual PR for posterity.

TL;DR: I believe the current state of this PR implements what is correct for backward compatibility

The issue at stake is the rules about what the effective security context for a container should be
if both `pod.Spec.SecurityContext` and `pod.Spec.Containers[x].SecurityContext` are set.  The
options discussed in #12823 were:

1. Generate the effective security context for a container by projecting
   `pod.Spec.Containers[x].SecurityContext` onto `pod.Spec.SecurityContext` in the API server and
   set the effective context on `pod.Spec.Containers[x].SecurityContext` in defaults or conversions
1. Generate the effective security context for a container by projecting
   `pod.Spec.Containers[x].SecurityContext` onto `pod.Spec.SecurityContext` at runtime in the
   kubelet
1. Have `pod.Spec.Containers[x].SecurityContext` completely override `pod.Spec.SecurityContext`

My own understanding of the problem was:

1. (1) is not allowable because v1.0.0 clients don't know how to deal with it.  V1.0.0 clients will
   not be able to send appliable patches or understand that the API server mutates the objects.
2. (2) is not allowable because v1.0.0 clients will not be able to reason correctly about what the
   effective security context is.  The behavior of the system will be changed by state which is not
   knowable by the client.
3. (3) was decided to be the only acceptable option for backward compatibility, since it lacks the
   deficiencies of (1) and (2):
   1.  v1.0.0 clients will always be able to reason about the correct effective security context of
       a container
   2.  v1.0.0 clients will be able to send accurate patches.

This PR is currently coded to implement (3).  I believe that @thockin is arguing for (2).  My
understanding is that (3) is the only viable option due to the rationale described above.

One thing I considered while writing the above:

We cannot make an easy decision about how to apply the rules for security context because we
relocated the host namespace fields on the internal API.  This means that a v1.0.0 pod that sets one
of the host namespace control bits in the pod spec will have a `pod.Spec.SecurityContext` set,
whether the pod author intended their to be one or not.  If we kept the host namespace fields on the
`PodSpec` in the V1 API we could apply a decision plane about how to handle these rules:

1.  If `pod.Spec.SecurityContext` is set, generate effective security context using (2)
2.  If `pod.Spec.SecurityContext` is *not* set, use the v1.0.0 semantics

...however, I believe that this would still present the problem element of (2) for v1.0.0 clients

For these reasons, I think what is coded currently is correct.

I think the fact that we're discussing this means that the proposal needs to articulate it more
clearly.  Once we get agreement here, I can modify the proposal in a follow-up PR.  For now, I am
going to hold off on updating the comments in the files on this PR as @thockin has asked until we
have sorted this out.

@thockin @bgrant0607 @smarterclayton
cc @pweil- @swagiaal

File: pkg/api/types.go (hunk 13 above)
Comment (pmorie) [inline pkg/api/types.go]: Going to hold off changing this until we recertify that we have agreement on what the semantics should be.

Comment (smarterclayton) [inline pkg/api/types.go]: I can be free on Monday.

**Comment by pmorie** [comment]
@swagiaal Suggested that I write up this example to illustrate where (2) falls down:

In the context of (2), say that we have a pod created by a new client:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  securityContext:
    runAsUser: 1001
    privileged: true
  containers:
    - name: test-container
      securityContext:
        runAsUser: 1002
```

A new client can reason about what `test-container`'s securityContext is going to be,
because it is aware of `PodSecurityContext`.  A new client will expect the security
context of `test-container` to be:

```yaml
runAsUser: 1002
privileged: true
```

A v1.0.0 client doesn't know about `PodSecurityContext` and so it can't make the right
inferences to determine the correct security context for `test-container`.  A v1.0.0
client will expect:

```yaml
runAsUser: 1002
```

This is what I mean by 'reasoning' in clients and it is where (2) above breaks old
clients.  You can vary the parameters by changing the `privileged` field in this
example and it will break down in the same way for v1.0.0 clients.

**Comment by bgrant0607** [comment]
@pmorie @smarterclayton @thockin 

Clients that previously performed writes or read-modify-write operations, but are oblivious to the new fields, must continue to work as they did before. 

I also want `kubectl apply` to work without custom logic for any fields, which means that we should avoid copying values to multiple fields during defaulting or conversion. Otherwise, the user would always have to specify both values, which would defeat the purpose of setting such values by default. This rules out (1).

However, the point of making default values explicit in the spec is so that system components and other components can infer desired state from the spec without needing to duplicate error-prone logic to reason about implicit default behavior -- see the original discussion in #1502.

Setting such default values works well for individual, orthogonal fields, but not in case of expansion of a single field into multiple fields nor in the case of fields that modify the behaviors of other fields. In particular, the latter implies that external observers that aren't fully up to date may not be able to fully comprehend the resulting behavior by looking at the spec. 

Though we should try to avoid instances of the latter, it appears that we can't avoid them entirely. Therefore, in some cases, we will need to add information in status to convey the rolled-up behavior. For instance, for scheduling constraints we should add a query that will return nodes that fit. For label selectors, we should return a label selector string that can be passed back to the server as the labelSelector query parameter to enable clients to enumerate matching resources without understanding the label selection semantics. Since such fields don't currently exist for SecurityContext, this doesn't help existing clients.

One could also view the pod-level SecurityContext as similar to LimitRange: constraints on the values that must be specified in the container-level SecurityContext.

I've concluded that we just need to relax compatibility rules for interpretation of spec fields of core resources in order to be able to evolve these APIs in a practical fashion.

Related discussion about evolution of label selectors and scheduling constraints:
https://github.com/kubernetes/kubernetes/issues/341#issuecomment-140809259

A variant of (3) would be to project container-level SecurityContext onto the pod-level SecurityContext. 

Options:
* overrides vs. constraints
* does pod-level override container-level, or vice versa?
* are the overrides for the whole SecurityContext or field by field?

Do we want any part of the container-level SecurityContext to remain forever, or do we expect it to go away entirely at some point? What end state would we like?

**Comment by thockin** [comment]
To Paul's example: I don't believe that `pod.security.privileged` should be a thing at all.  For other existing fields, they always had DTRT semantics.  RunAsNonRoot is a little weird - what if I set that to false - we should change that to a pointer type.

**Comment by pmorie** [comment]
@bgrant0607 

> One could also view the pod-level SecurityContext as similar to LimitRange: constraints on the values that must be specified in the container-level SecurityContext.

That's not what we intent `PodSecurityContext` to be -- constraints are separate and described in #7893

**Comment by pmorie** [comment]
@thockin 

> To Paul's example: I don't believe that pod.security.privileged should be a thing at all.

I would be okay subtracting `Privileged` from `PodSecurityContext` unless someone can make a strong case for needing to make all containers in a pod privileged without saying so on each container.

**Comment by pmorie** [comment]
> I would be okay subtracting Privileged from PodSecurityContext unless someone can make a strong case for needing to make all containers in a pod privileged without saying so on each container.

Same thing goes for Capabilities.

**Comment by bgrant0607** [comment]
I just noticed that documentation of SELinuxOptions fields incorrectly reference docs/user-guide/labels.md:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L2503

**Comment by bgrant0607** [comment]
One other option to consider: we could make setting container-level SecurityContext fields illegal when overlapping pod-level fields are set, and vice versa. Not my favorite option, and could break someone, but it would at least result in understandable behavior rather than silently produce unexpected behavior. RunAsNonRoot would need to be converted to a pointer in order to distinguish unset from golang default, as discussed in #15566.

**Comment by bgrant0607** [comment]
Filed #15879 re. updating api conventions.

**Comment by pmorie** [comment]
For posterity:

We had a mismatch.  My action plan is:

1. Refactor to (2)
2. Remove `Capabilities` and `Privileged` from `PodSecurityContext`
3. Convert `RunAsNonRoot` to a pointer
4. Fix comments on existing `SecurityContext` fields

**Comment by pmorie** [comment]
@thockin @smarterclayton @bgrant0607 PTAL

File: pkg/api/types.go [hunk 15]
```diff
@@ -789,7 +789,8 @@ type Container struct {
 	TerminationMessagePath string `json:"terminationMessagePath,omitempty"`
 	// Required: Policy for pulling images for this container
 	ImagePullPolicy PullPolicy `json:"imagePullPolicy"`
-	// Optional: SecurityContext defines the security options the pod should be run with
+	// Optional: SecurityContext defines the security options the container should be run with.
+	// If set, the fields of SecurityContext override the equivalent fields of PodSecurityContext.
 	SecurityContext *SecurityContext `json:"securityContext,omitempty"`
```
Comment (thockin) [inline pkg/api/types.go]: we could/should rename SecurityContext to ContainerSecurityContext - different PR

File: pkg/api/types.go [hunk 16]
```diff
@@ -2031,41 +2053,45 @@ type ComponentStatusList struct {
...
-	// Run the container in privileged mode
+	// Run container in privileged mode.
+	// Processes in privileged containers are essentially equivalent to root on the host.
+	// Defaults to false.
 	Privileged *bool `json:"privileged,omitempty"`
-
-	// SELinuxOptions are the labels to be applied to the container
-	// and volumes
+	// The SELinux context to be applied to the container.
+	// If unspecified, the container runtime will allocate a random SELinux context for each
```
Comment (thockin) [inline pkg/api/types.go]: "if unspecified" is said twice.

**Comment by thockin** [comment]
soooooo close, I'm gonna LGTM this and say you can fix your typo and tag it lgtm yourself :)

**Comment by pmorie** [comment]
I have a couple nits I'm fixing from rebasing onto #14991 that I'm still ironing out about this.

**Comment by pmorie** [comment]
This should be ready for LGTM.  Waiting to see what CI tells us before re-adding.

File: pkg/securitycontext/provider.go [hunk 17]
```diff
@@ -38,17 +38,20 @@ type SimpleSecurityContextProvider struct{}
...
-		config.User = strconv.FormatInt(*container.SecurityContext.RunAsUser, 10)
+	if effectiveSC.RunAsUser != nil {
+		config.User = strconv.Itoa(int(*effectiveSC.RunAsUser))
 	}
 }
 
 // ModifyHostConfig is called before the Docker runContainer call.
 // The security context provider can make changes to the HostConfig, affecting
 // security options, whether the container is privileged, volume binds, etc.
+// An error is returned if it's not possible to secure the container as requested
```
Comment (pweil-) [inline pkg/securitycontext/provider.go]: remove these two lines from the merge

**Comment by pmorie** [comment]
By the power vested in me by @thockin in https://github.com/kubernetes/kubernetes/pull/14705#issuecomment-149783453, I declare this PR to be LGTM

**Comment by pmorie** [comment]
C'mon shippable :game_die: :game_die:

**Comment by pmorie** [comment]
Noticed one last bug and had to slip to the back of the shippable queue.  I'll be babysitting this PR tonight. :baby: :baby_bottle:

**Comment by pmorie** [comment]
@thockin It sounds like you think this should be merged.  Vote with your finger at hit the button? :ballot_box_with_check:

**Comment by pmorie** [comment]
Shippable build timed out.

I'm merging as per @thockin's request here: https://github.com/kubernetes/kubernetes/pull/15323#issuecomment-150055456

Let it be known that I gave shippable and merge-bot a chance and that I consider self-merge to be a last option.  Here goes nothing. :pray: :pray: :pray:

**Comment by piosz** [comment]
This PR broke hack/verify-api-reference-docs.sh check. Fixed in #16092.
